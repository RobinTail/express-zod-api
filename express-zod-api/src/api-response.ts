import { z } from "zod";
import { isSchema } from "./common-helpers";
import { contentTypes } from "./content-type";

export const defaultStatusCodes = {
  positive: 200,
  negative: 400,
} satisfies Record<string, number>;

export type ResponseVariant = keyof typeof defaultStatusCodes;
export const responseVariants = Object.keys(
  defaultStatusCodes,
) as ResponseVariant[];

/** @public this is the user facing configuration */
export class ApiResponse<S extends z.ZodType> {
  readonly #schema: S;
  readonly #statusCodes?: [number, ...number[]];
  readonly #mimeTypes?: [string, ...string[]] | null;

  public constructor(
    subject:
      | S // shorthand for { schema: S }, default status codes and MIME types applied
      | {
          schema: S;
          /** @default 200 for a positive and 400 for a negative response */
          statusCode?: number | [number, ...number[]];
          /**
           * @example null is for no content, such as 204 and 302
           * @default "application/json"
           * */
          mimeType?: string | [string, ...string[]] | null;
        },
  ) {
    const { schema, statusCode, mimeType } = isSchema(subject)
      ? { schema: subject }
      : subject;
    this.#schema = schema;
    this.#statusCodes =
      typeof statusCode === "number" ? [statusCode] : statusCode;
    this.#mimeTypes = typeof mimeType === "string" ? [mimeType] : mimeType;
  }

  public normalize(fbStatus: number) {
    return {
      schema: this.#schema,
      statusCodes: this.#statusCodes ?? [fbStatus],
      mimeTypes:
        this.#mimeTypes === undefined ? [contentTypes.json] : this.#mimeTypes,
    };
  }
}

export type NormalizedResponse = ReturnType<
  ApiResponse<z.ZodType>["normalize"]
>;
