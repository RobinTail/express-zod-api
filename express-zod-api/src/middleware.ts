import { NextFunction, Request, Response } from "express";
import { z } from "zod";
import { EmptySchema, FlatObject } from "./common-helpers";
import { InputValidationError } from "./errors";
import { IOSchema } from "./io-schema";
import { LogicalContainer } from "./logical-container";
import { Security } from "./security";
import { ActualLogger } from "./logger-helpers";

type Handler<IN, OPT, OUT> = (params: {
  input: IN;
  options: OPT;
  request: Request;
  response: Response;
  logger: ActualLogger;
}) => Promise<OUT>;

export abstract class AbstractMiddleware {
  public abstract getSecurity(): LogicalContainer<Security> | undefined;
  public abstract getSchema(): IOSchema<"strip">;
  public abstract execute(params: {
    input: unknown;
    options: FlatObject;
    request: Request;
    response: Response;
    logger: ActualLogger;
  }): Promise<FlatObject>;
}

export class Middleware<
  OPT extends FlatObject,
  OUT extends FlatObject,
  SCO extends string,
  IN extends IOSchema<"strip"> = EmptySchema,
> extends AbstractMiddleware {
  readonly #schema: IN;
  readonly #security?: LogicalContainer<
    Security<Extract<keyof z.input<IN>, string>, SCO>
  >;
  readonly #handler: Handler<z.output<IN>, OPT, OUT>;

  constructor({
    input = z.object({}) as IN,
    security,
    handler,
  }: {
    input?: IN;
    security?: LogicalContainer<
      Security<Extract<keyof z.input<IN>, string>, SCO>
    >;
    handler: Handler<z.output<IN>, OPT, OUT>;
  }) {
    super();
    this.#schema = input;
    this.#security = security;
    this.#handler = handler;
  }

  public override getSecurity() {
    return this.#security;
  }

  public override getSchema() {
    return this.#schema;
  }

  /** @throws InputValidationError */
  public override async execute({
    input,
    ...rest
  }: {
    input: unknown;
    options: OPT;
    request: Request;
    response: Response;
    logger: ActualLogger;
  }) {
    try {
      const validInput = (await this.#schema.parseAsync(input)) as z.output<IN>;
      return this.#handler({ ...rest, input: validInput });
    } catch (e) {
      throw e instanceof z.ZodError ? new InputValidationError(e) : e;
    }
  }
}

export class ExpressMiddleware<
  R extends Request,
  S extends Response,
  OUT extends FlatObject,
> extends Middleware<FlatObject, OUT, string> {
  constructor(
    nativeMw: (
      request: R,
      response: S,
      next: NextFunction,
    ) => void | Promise<void>,
    {
      provider = () => ({}) as OUT,
      transformer = (err: Error) => err,
    }: {
      provider?: (request: R, response: S) => OUT | Promise<OUT>;
      transformer?: (err: Error) => Error;
    } = {},
  ) {
    super({
      handler: async ({ request, response }) =>
        new Promise<OUT>((resolve, reject) => {
          const next = (err?: unknown) => {
            if (err && err instanceof Error) return reject(transformer(err));
            resolve(provider(request as R, response as S));
          };
          nativeMw(request as R, response as S, next)?.catch(next);
        }),
    });
  }
}
