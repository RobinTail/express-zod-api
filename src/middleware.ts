import { Request, Response } from "express";
import { HttpError } from "http-errors";
import { Logger } from "winston";
import { z } from "zod";
import { FlatObject, IOSchema } from "./common-helpers";
import { LogicalContainer } from "./logical-container";
import { Security } from "./security";

interface MiddlewareParams<IN, OPT> {
  input: IN;
  options: OPT;
  request: Request;
  response: Response;
  logger: Logger;
}

type Middleware<IN, OPT, OUT> = (
  params: MiddlewareParams<IN, OPT>
) => Promise<OUT>;

export interface MiddlewareCreationProps<
  IN extends IOSchema<"strip">,
  OPT,
  OUT extends FlatObject
> {
  input: IN;
  security?: LogicalContainer<Security<keyof z.input<IN> & string>>;
  middleware: Middleware<z.output<IN>, OPT, OUT>;
}

export interface MiddlewareDefinition<
  IN extends IOSchema<"strip">,
  OPT,
  OUT extends FlatObject
> extends MiddlewareCreationProps<IN, OPT, OUT> {
  type: "proprietary" | "express";
}

export type AnyMiddlewareDef = MiddlewareDefinition<any, any, any>;

export const createMiddleware = <
  IN extends IOSchema<"strip">,
  OPT,
  OUT extends FlatObject
>(
  props: MiddlewareCreationProps<IN, OPT, OUT>
): MiddlewareDefinition<IN, OPT, OUT> => ({
  ...props,
  type: "proprietary",
});

export type ExpressMiddleware<R extends Request, S extends Response> = (
  request: R,
  response: S,
  next: (error?: any) => void
) => void | Promise<void>;

export interface ExpressMiddlewareFeatures<
  R extends Request,
  S extends Response,
  OUT extends FlatObject
> {
  provider?: (request: R, response: S) => OUT | Promise<OUT>;
  transformer?: (err: Error) => HttpError | Error;
}
