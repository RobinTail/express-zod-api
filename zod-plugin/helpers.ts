import type { z } from "zod";
import { unpack } from "./packer";

export const getBrand = (subject: z.core.$ZodType) => {
  const { brand } = unpack(subject) || {};
  if (
    typeof brand === "symbol" ||
    typeof brand === "string" ||
    typeof brand === "number"
  )
    return brand;
  return undefined;
};
