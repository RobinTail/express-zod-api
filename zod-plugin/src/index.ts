import "./runtime.ts"; // side effects here
export { pack, unpack } from "./packer.ts";
export { getBrand } from "./brand.ts";
