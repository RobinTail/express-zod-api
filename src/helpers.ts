import {Request} from 'express';
import {z} from 'zod';
import {MiddlewareDefinition} from './middleware';

export type FlatObject = Record<string, any>;
export type ObjectSchema = z.AnyZodObject;

export type Merge<A extends ObjectSchema, B extends ObjectSchema | any> = z.ZodObject<
  // eslint-disable-next-line @typescript-eslint/ban-types
  A['shape'] & (B extends ObjectSchema ? B['shape'] : {}),
  A['_unknownKeys'],
  A['_catchall']
>;

export function combineEndpointAndMiddlewareInputSchemas<IN extends ObjectSchema, mIN>(
  input: IN,
  middlewares: MiddlewareDefinition<any, any, any>[]
): Merge<IN, mIN> {
  if (middlewares.length === 0) {
    return input as any as Merge<IN, mIN>;
  }
  return middlewares
    .map((middleware) => middleware.input)
    .reduce((carry: ObjectSchema, schema) => carry.merge(schema))
    .merge(input) as Merge<IN, mIN>;
}

export function getInitialInput(request: Request): any {
  switch (request.method) {
    case 'POST':
    case 'PUT':
    case 'PATCH':
      return request.body;
    case 'GET':
      return request.query;
    case 'DELETE': // _may_ have body
    default:
      return {...request.query, ...request.body};
  }
}
