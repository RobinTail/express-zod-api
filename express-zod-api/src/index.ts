import "@express-zod-api/zod-plugin"; // side effects here
export { createConfig } from "./config-type.ts";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory.ts";
export { getMessageFromError } from "./common-helpers.ts";
export { ensureHttpError } from "./result-helpers.ts";
export { BuiltinLogger } from "./builtin-logger.ts";
export { Middleware } from "./middleware.ts";
export {
  ResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler.ts";
export { DependsOnMethod } from "./depends-on-method.ts";
export { ServeStatic } from "./serve-static.ts";
export { createServer, attachRouting } from "./server.ts";
export { Documentation } from "./documentation.ts";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors.ts";
export { testEndpoint, testMiddleware } from "./testing.ts";
export { Integration } from "./integration.ts";
export { EventStreamFactory } from "./sse.ts";

export { ez } from "./proprietary-schemas.ts";

// Convenience types
export type { Depicter } from "./documentation-helpers.ts";
export type { Producer } from "./zts-helpers.ts";

// Interfaces exposed for augmentation
export type { LoggerOverrides } from "./logger-helpers.ts";
export type { TagOverrides } from "./common-helpers.ts";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
import type {} from "qs"; // fixes TS2742 for attachRouting
export type { Routing } from "./routing.ts";
export type { FlatObject } from "./common-helpers.ts";
export type { Method } from "./method.ts";
export type { IOSchema } from "./io-schema.ts";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type.ts";
export type { ApiResponse } from "./api-response.ts";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  HeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security.ts";
