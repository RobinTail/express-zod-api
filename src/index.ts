import "./zod-plugin";

export { createConfig } from "./config-type";
export { AbstractEndpoint } from "./endpoint";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory";
export {
  getExamples,
  getMessageFromError,
  getStatusCodeFromError,
} from "./common-helpers";
export { createLogger } from "./logger";
export { createMiddleware } from "./middleware";
export {
  createResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler";
export { DependsOnMethod } from "./depends-on-method";
export { ServeStatic } from "./serve-static";
export { createServer, attachRouting } from "./server";
export { Documentation } from "./documentation";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors";
export { testEndpoint } from "./testing";
export { Integration } from "./integration";

export { ez } from "./proprietary-schemas";

// Convenience types
export type { Depicter } from "./documentation-helpers";
export type { Producer } from "./zts-helpers";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
export type { MockOverrides } from "./testing";
export type { Routing } from "./routing";
export type { LoggerOverrides } from "./logger";
export type { FlatObject } from "./common-helpers";
export type { Method } from "./method";
export type { IOSchema } from "./io-schema";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type";
export type { MiddlewareDefinition } from "./middleware";
export type { ResultHandlerDefinition } from "./result-handler";
export type { ApiResponse } from "./api-response";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  CustomHeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security";
