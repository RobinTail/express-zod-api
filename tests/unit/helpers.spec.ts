import {expectType} from 'tsd';
import {
  combineEndpointAndMiddlewareInputSchemas,
  extractObjectSchema,
  getInitialInput,
  getMessageFromError,
  getStatusCodeFromError,
  isLoggerConfig,
  OutputMarker
} from '../../src/helpers';
import {createMiddleware, z, createHttpError, markOutput, createApiResponse} from '../../src';
import {Request} from 'express';
import {MiddlewareDefinition} from '../../src/middleware';
import {serializeSchemaForTest} from '../helpers';

describe('Helpers', () => {
  describe('combineEndpointAndMiddlewareInputSchemas()', () => {
    test('Should merge input object schemas', () => {
      const middlewares = [
        createMiddleware({
          input: z.object({
            one: z.string()
          }),
          middleware: jest.fn()
        }),
        createMiddleware({
          input: z.object({
            two: z.number()
          }),
          middleware: jest.fn()
        }),
        createMiddleware({
          input: z.object({
            three: z.null()
          }),
          middleware: jest.fn()
        }),
      ] as MiddlewareDefinition<any, any, any>[];
      const endpointInput = z.object({
        four: z.boolean()
      });
      const result = combineEndpointAndMiddlewareInputSchemas(endpointInput, middlewares);
      expect(result).toBeInstanceOf(z.ZodObject);
      expect(serializeSchemaForTest(result)).toMatchSnapshot();
    });

    test('Should merge union object schemas', () => {
      const middlewares = [
        createMiddleware({
          input: z.object({
            one: z.string()
          }).or(z.object({
            two: z.number()
          })),
          middleware: jest.fn()
        }),
        createMiddleware({
          input: z.object({
            three: z.null()
          }).or(z.object({
            four: z.boolean()
          })),
          middleware: jest.fn()
        }),
      ] as MiddlewareDefinition<any, any, any>[];
      const endpointInput = z.object({
        five: z.string()
      }).or(z.object({
        six: z.number()
      }));
      const result = combineEndpointAndMiddlewareInputSchemas(endpointInput, middlewares);
      expect(result).toBeInstanceOf(z.ZodObject);
      expect(serializeSchemaForTest(result)).toMatchSnapshot();
    });

    test('Should merge intersection object schemas', () => {
      const middlewares = [
        createMiddleware({
          input: z.object({
            one: z.string()
          }).and(z.object({
            two: z.number()
          })),
          middleware: jest.fn()
        }),
        createMiddleware({
          input: z.object({
            three: z.null()
          }).and(z.object({
            four: z.boolean()
          })),
          middleware: jest.fn()
        }),
      ] as MiddlewareDefinition<any, any, any>[];
      const endpointInput = z.object({
        five: z.string()
      }).and(z.object({
        six: z.number()
      }));
      const result = combineEndpointAndMiddlewareInputSchemas(endpointInput, middlewares);
      expect(result).toBeInstanceOf(z.ZodObject);
      expect(serializeSchemaForTest(result)).toMatchSnapshot();
    });

    test('Should merge mixed object schemas', () => {
      const middlewares = [
        createMiddleware({
          input: z.object({
            one: z.string()
          }).and(z.object({
            two: z.number()
          })),
          middleware: jest.fn()
        }),
        createMiddleware({
          input: z.object({
            three: z.null()
          }).or(z.object({
            four: z.boolean()
          })),
          middleware: jest.fn()
        }),
      ] as MiddlewareDefinition<any, any, any>[];
      const endpointInput = z.object({
        five: z.string()
      });
      const result = combineEndpointAndMiddlewareInputSchemas(endpointInput, middlewares);
      expect(result).toBeInstanceOf(z.ZodObject);
      expect(serializeSchemaForTest(result)).toMatchSnapshot();
    });
  });
  
  describe('getInitialInput()', () => {
    test('should return body for POST, PUT and PATCH requests', () => {
      expect(getInitialInput({
        body: 'body',
        method: 'POST'
      } as Request)).toEqual('body');
      expect(getInitialInput({
        body: 'body',
        method: 'PUT'
      } as Request)).toEqual('body');
      expect(getInitialInput({
        body: 'body',
        method: 'PATCH'
      } as Request)).toEqual('body');
    });
    test('should return query for GET requests', () => {
      expect(getInitialInput({
        query: 'query',
        method: 'GET'
      } as unknown as Request)).toEqual('query');
    });
    test('should return both body and query for DELETE and unknown requests', () => {
      expect(getInitialInput({
        query: { a: 'query' },
        body: {b: 'body'},
        method: 'DELETE'
      } as unknown as Request)).toEqual({
        a: 'query',
        b: 'body'
      });
    });
  });

  describe('isLoggerConfig()', () => {
    test('Should identify the valid logger config', () => {
      expect(isLoggerConfig({
        level: 'debug',
        color: true,
      })).toBeTruthy();
    });
    test('Should reject the object with invalid properties', () => {
      expect(isLoggerConfig({
        level: 'something',
        color: true,
      })).toBeFalsy();
      expect(isLoggerConfig({
        level: 'debug',
        color: null,
      })).toBeFalsy();
    });
    test('Should reject the object with missing properties', () => {
      expect(isLoggerConfig({
        level: 'something',
      })).toBeFalsy();
      expect(isLoggerConfig({
        color: null,
      })).toBeFalsy();
    });
    test('Should reject non-objects', () => {
      expect(isLoggerConfig([1,2,3])).toBeFalsy();
      expect(isLoggerConfig('something')).toBeFalsy();
    });
  });

  describe('extractObjectSchema()', () => {
    test('should pass the object schema through', () => {
      const subject = extractObjectSchema(z.object({
        one: z.string()
      }));
      expect(subject).toBeInstanceOf(z.ZodObject);
      expect(serializeSchemaForTest(subject)).toMatchSnapshot();
    });

    test('should return object schema for the union of object schemas', () => {
      const subject = extractObjectSchema(z.object({
        one: z.string()
      }).or(z.object({
        two: z.number()
      })));
      expect(subject).toBeInstanceOf(z.ZodObject);
      expect(serializeSchemaForTest(subject)).toMatchSnapshot();
    });

    test('should return object schema for the intersection of object schemas', () => {
      const subject = extractObjectSchema(z.object({
        one: z.string()
      }).and(z.object({
        two: z.number()
      })));
      expect(subject).toBeInstanceOf(z.ZodObject);
      expect(serializeSchemaForTest(subject)).toMatchSnapshot();
    });
  });

  describe('getMessageFromError()', () => {
    test('should compile a string from ZodError', () => {
      const error = new z.ZodError([
        {
          code: 'invalid_type',
          path: ['user', 'id'],
          message: 'expected number, got string',
          expected: 'number',
          received: 'string'
        },
        {
          code: 'invalid_type',
          path: ['user', 'name'],
          message: 'expected string, got number',
          expected: 'string',
          received: 'number'
        }
      ]);
      expect(getMessageFromError(error)).toMatchSnapshot();
    });

    test('should pass message from other error types', () => {
      expect(getMessageFromError(
        createHttpError(502, 'something went wrong'))
      ).toMatchSnapshot();
      expect(getMessageFromError(
        new Error('something went wrong'))
      ).toMatchSnapshot();
    });
  });

  describe('getStatusCodeFromError()', () => {
    test('should get status code from HttpError', () => {
      expect(getStatusCodeFromError(createHttpError(403, 'Access denied')))
        .toEqual(403);
    });

    test('should return 400 for ZodError', () => {
      const error = new z.ZodError([
        {
          code: 'invalid_type',
          path: ['user', 'id'],
          message: 'expected number, got string',
          expected: 'number',
          received: 'string'
        }
      ]);
      expect(getStatusCodeFromError(error)).toEqual(400);
    });

    test('should return 500 for other errors', () => {
      expect(getStatusCodeFromError(new Error('something went wrong')))
        .toEqual(500);
    });
  });

  describe('markOutput()', () => {
    test('should change the type of schema', () => {
      const output = z.object({});
      expect(markOutput(output)).toEqual(output);
      expectType<OutputMarker>(markOutput(output));
    });
  });

  describe('createApiResponse()', () => {
    test('should accept an array of mime types', () => {
      const output = z.object({});
      expect(createApiResponse(output, ['something', 'anything'])).toEqual({
        schema: output,
        mimeTypes: ['something', 'anything']
      });
    });

    test('should accept a single mime type', () => {
      const output = z.object({});
      expect(createApiResponse(output, 'something')).toEqual({
        schema: output,
        mimeTypes: ['something']
      });
    });

    test('should assume json mime type by default', () => {
      const output = z.object({});
      expect(createApiResponse(output, )).toEqual({
        schema: output,
        mimeTypes: ['application/json']
      });
    });
  });
});
