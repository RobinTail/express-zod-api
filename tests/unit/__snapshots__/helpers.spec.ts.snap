// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helpers combineEndpointAndMiddlewareInputSchemas() Should merge input object schemas 1`] = `
Object {
  "_type": "ZodObject",
  "shape": Object {
    "four": Object {
      "_type": "ZodBoolean",
    },
    "one": Object {
      "_type": "ZodString",
    },
    "three": Object {
      "_type": "ZodNull",
    },
    "two": Object {
      "_type": "ZodNumber",
    },
  },
}
`;

exports[`Helpers combineEndpointAndMiddlewareInputSchemas() Should merge intersection object schemas 1`] = `
Object {
  "_type": "ZodObject",
  "shape": Object {
    "five": Object {
      "_type": "ZodString",
    },
    "four": Object {
      "_type": "ZodBoolean",
    },
    "one": Object {
      "_type": "ZodString",
    },
    "six": Object {
      "_type": "ZodNumber",
    },
    "three": Object {
      "_type": "ZodNull",
    },
    "two": Object {
      "_type": "ZodNumber",
    },
  },
}
`;

exports[`Helpers combineEndpointAndMiddlewareInputSchemas() Should merge mixed object schemas 1`] = `
Object {
  "_type": "ZodObject",
  "shape": Object {
    "five": Object {
      "_type": "ZodString",
    },
    "four": Object {
      "_type": "ZodOptional",
      "value": Object {
        "_type": "ZodBoolean",
      },
    },
    "one": Object {
      "_type": "ZodString",
    },
    "three": Object {
      "_type": "ZodOptional",
      "value": Object {
        "_type": "ZodNull",
      },
    },
    "two": Object {
      "_type": "ZodNumber",
    },
  },
}
`;

exports[`Helpers combineEndpointAndMiddlewareInputSchemas() Should merge union object schemas 1`] = `
Object {
  "_type": "ZodObject",
  "shape": Object {
    "five": Object {
      "_type": "ZodOptional",
      "value": Object {
        "_type": "ZodString",
      },
    },
    "four": Object {
      "_type": "ZodOptional",
      "value": Object {
        "_type": "ZodBoolean",
      },
    },
    "one": Object {
      "_type": "ZodOptional",
      "value": Object {
        "_type": "ZodString",
      },
    },
    "six": Object {
      "_type": "ZodOptional",
      "value": Object {
        "_type": "ZodNumber",
      },
    },
    "three": Object {
      "_type": "ZodOptional",
      "value": Object {
        "_type": "ZodNull",
      },
    },
    "two": Object {
      "_type": "ZodOptional",
      "value": Object {
        "_type": "ZodNumber",
      },
    },
  },
}
`;

exports[`Helpers extractObjectSchema() should pass the object schema through 1`] = `
Object {
  "_type": "ZodObject",
  "shape": Object {
    "one": Object {
      "_type": "ZodString",
    },
  },
}
`;

exports[`Helpers extractObjectSchema() should return object schema for the intersection of object schemas 1`] = `
Object {
  "_type": "ZodObject",
  "shape": Object {
    "one": Object {
      "_type": "ZodString",
    },
    "two": Object {
      "_type": "ZodNumber",
    },
  },
}
`;

exports[`Helpers extractObjectSchema() should return object schema for the union of object schemas 1`] = `
Object {
  "_type": "ZodObject",
  "shape": Object {
    "one": Object {
      "_type": "ZodOptional",
      "value": Object {
        "_type": "ZodString",
      },
    },
    "two": Object {
      "_type": "ZodOptional",
      "value": Object {
        "_type": "ZodNumber",
      },
    },
  },
}
`;

exports[`Helpers getMessageFromError() should compile a string from ZodError 1`] = `"user/id: expected number, got string; user/name: expected string, got number"`;

exports[`Helpers getMessageFromError() should pass message from other error types 1`] = `"something went wrong"`;

exports[`Helpers getMessageFromError() should pass message from other error types 2`] = `"something went wrong"`;
